src/lib/operators/export_binary.cpp

```diff
@@ -171,90 +149,70 @@ void ExportBinary::_write_header(const Table& table, std::ofstream& ofstream) {
-
-  const auto context = std::make_shared<ExportContext>(ofstream);
-
   export_value(ofstream, static_cast<ChunkOffset>(chunk->size()));
 
   // Iterating over all segments of this chunk and exporting them
   for (ColumnID column_id{0}; column_id < chunk->column_count(); column_id++) {
-    auto visitor =
-        make_unique_by_data_type<AbstractSegmentVisitor, ExportBinaryVisitor>(table.column_data_type(column_id));
-    resolve_data_and_segment_type(*chunk->get_segment(column_id),
-                                  [&](const auto data_type_t, const auto& resolved_segment) {
-                                    visitor->handle_segment(resolved_segment, context);
-                                  });
+    resolve_data_and_segment_type(
+        *chunk->get_segment(column_id),
+        [&](const auto data_type_t, const auto& resolved_segment) { _write_segment(resolved_segment, ofstream); });
   }
 }
 
-template <typename T>
-void ExportBinary::ExportBinaryVisitor<T>::handle_segment(const BaseValueSegment& base_segment,
-                                                          std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<ExportContext>(base_context);
-  const auto& segment = static_cast<const ValueSegment<T>&>(base_segment);
+void ExportBinary::_write_segment(const BaseEncodedSegment& base_segment, std::ofstream& ofstream) {
+  Fail("Binary export for segment type is not supported yet.");
+}
 
-  export_value(context->ofstream, BinarySegmentType::value_segment);
+template <typename T>
+void ExportBinary::_write_segment(const ValueSegment<T>& segment, std::ofstream& ofstream) {
+  export_value(ofstream, EncodingType::Unencoded);
 
   if (segment.is_nullable()) {
-    export_values(context->ofstream, segment.null_values());
+    export_values(ofstream, segment.null_values());
   }
 
-  export_values(context->ofstream, segment.values());
+  export_values(ofstream, segment.values());
 }
 
-template <typename T>
-void ExportBinary::ExportBinaryVisitor<T>::handle_segment(const ReferenceSegment& ref_segment,
-                                                          std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<ExportContext>(base_context);
-
+void ExportBinary::_write_segment(const ReferenceSegment& ref_segment, std::ofstream& ofstream) {
   // We materialize reference segments and save them as value segments
-  export_value(context->ofstream, BinarySegmentType::value_segment);
+  export_value(ofstream, EncodingType::Unencoded);
 
-  // Unfortunately, we have to iterate over all values of the reference segment
-  // to materialize its contents. Then we can write them to the file
-  for (ChunkOffset row = 0; row < ref_segment.size(); ++row) {
-    export_value(context->ofstream, boost::get<T>(ref_segment[row]));
-  }
-}
+  if (ref_segment.data_type() == DataType::String) {
+    if (ref_segment.size() == 0) return;
```

Comment:
Not your code, but unless we have a good reason not to, we should write out `reference_segment`